File: go/cpu/matmul.go
package main

import (
	"math"

	"github.com/mohamedmenasy/go-vs-rust-benchmarks/go/internal/common"
	"github.com/mohamedmenasy/go-vs-rust-benchmarks/go/internal/harness"
)

// matmul: C = A x B for n x n row-major float64 matrices, naive i-k-j loop
// order. Two formulations of the same arithmetic:
//
//   - matmul      (baseline): flat indexing c[i*n+j] += a[i*n+k] * b[k*n+j]
//   - matmul-rows (idiomatic): row sub-slices, letting each compiler drop
//     bounds checks (and, where it can, vectorize the inner loop)
//
// Each C element accumulates over k in the same order in both languages and
// neither compiler emits FMA at the baseline ISA, so results are bit-identical.
type matmul struct {
	n       int
	a, b, c []float64
	rows    bool
}

func (w *matmul) Prepare() { clear(w.c) }

func (w *matmul) Run() uint64 {
	if w.rows {
		matmulRows(w.n, w.a, w.b, w.c)
	} else {
		matmulFlat(w.n, w.a, w.b, w.c)
	}
	return math.Float64bits(w.c[0]) ^ math.Float64bits(w.c[w.n*w.n-1])
}

func (w *matmul) Check() uint64 {
	d := common.NewDigest()
	for _, x := range w.c {
		d.AddF64(x)
	}
	return d.Sum()
}

func matmulFlat(n int, a, b, c []float64) {
	for i := 0; i < n; i++ {
		for k := 0; k < n; k++ {
			aik := a[i*n+k]
			for j := 0; j < n; j++ {
				c[i*n+j] += aik * b[k*n+j]
			}
		}
	}
}

func matmulRows(n int, a, b, c []float64) {
	for i := 0; i < n; i++ {
		ci := c[i*n : i*n+n]
		ai := a[i*n : i*n+n]
		for k, aik := range ai {
			bk := b[k*n : k*n+n]
			bk = bk[:len(ci)] // bounds-check-elimination hint
			for j := range ci {
				ci[j] += aik * bk[j]
			}
		}
	}
}

func newMatmul(p harness.Params, rows bool) (harness.Instance, harness.Work, error) {
	n := int(p.Int("n", 512))
	r := common.NewSplitMix64(uint64(p.Int("seed", 3)))
	a := make([]float64, n*n)
	b := make([]float64, n*n)
	for i := range a {
		a[i] = r.Float64()
	}
	for i := range b {
		b[i] = r.Float64()
	}
	flops := 2 * float64(n) * float64(n) * float64(n)
	return &matmul{n: n, a: a, b: b, c: make([]float64, n*n), rows: rows},
		harness.Work{Unit: "flops", PerRun: flops}, nil
}

func setupMatmul(p harness.Params) (harness.Instance, harness.Work, error) {
	return newMatmul(p, false)
}

func setupMatmulRows(p harness.Params) (harness.Instance, harness.Work, error) {
	return newMatmul(p, true)
}

File: rust/cpu/src/matmul.rs
//! C = A x B for n x n row-major f64 matrices, naive i-k-j loop order.
//!
//! * `matmul` (baseline): flat indexing `c[i*n+j] += a[i*n+k] * b[k*n+j]`
//! * `matmul-rows` (idiomatic): row sub-slices zipped with iterators, letting
//!   the compiler drop bounds checks and vectorize the inner loop.
//!
//! Each C element accumulates over k in the same order as go/cpu/matmul.go
//! and no FMA contraction happens at the baseline ISA: results are
//! bit-identical.

use bench_common::{Digest, Instance, Params, SplitMix64, Work};

struct Matmul {
    n: usize,
    a: Vec<f64>,
    b: Vec<f64>,
    c: Vec<f64>,
    rows: bool,
}

fn matmul_flat(n: usize, a: &[f64], b: &[f64], c: &mut [f64]) {
    for i in 0..n {
        for k in 0..n {
            let aik = a[i * n + k];
            for j in 0..n {
                c[i * n + j] += aik * b[k * n + j];
            }
        }
    }
}

fn matmul_rows(n: usize, a: &[f64], b: &[f64], c: &mut [f64]) {
    for (ci, ai) in c.chunks_exact_mut(n).zip(a.chunks_exact(n)) {
        for (&aik, bk) in ai.iter().zip(b.chunks_exact(n)) {
            for (cij, &bkj) in ci.iter_mut().zip(bk) {
                *cij += aik * bkj;
            }
        }
    }
}

impl Instance for Matmul {
    fn prepare(&mut self) {
        self.c.fill(0.0);
    }
    fn run(&mut self) -> u64 {
        if self.rows {
            matmul_rows(self.n, &self.a, &self.b, &mut self.c);
        } else {
            matmul_flat(self.n, &self.a, &self.b, &mut self.c);
        }
        self.c[0].to_bits() ^ self.c[self.n * self.n - 1].to_bits()
    }
    fn has_check(&self) -> bool {
        true
    }
    fn check(&mut self) -> u64 {
        let mut d = Digest::new();
        for &x in &self.c {
            d.add_f64(x);
        }
        d.sum()
    }
}

fn setup(p: &Params, rows: bool) -> Result<(Box<dyn Instance>, Work), String> {
    let n = p.int("n", 512) as usize;
    let mut r = SplitMix64::new(p.int("seed", 3) as u64);
    let a: Vec<f64> = (0..n * n).map(|_| r.float64()).collect();
    let b: Vec<f64> = (0..n * n).map(|_| r.float64()).collect();
    let flops = 2.0 * (n as f64).powi(3);
    Ok((
        Box::new(Matmul { n, a, b, c: vec![0.0; n * n], rows }),
        Work { unit: "flops", per_run: flops, input_bytes: 0 },
    ))
}

pub fn setup_flat(p: &Params) -> Result<(Box<dyn Instance>, Work), String> {
    setup(p, false)
}

pub fn setup_rows(p: &Params) -> Result<(Box<dyn Instance>, Work), String> {
    setup(p, true)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testutil::{golden_cpu, params, run_once};
    use bench_common::golden_hex;

    #[test]
    fn matches_golden() {
        let c = &golden_cpu()["matmul"];
        for s in [setup_flat as bench_common::harness::SetupFn, setup_rows] {
            let (mut inst, _) =
                s(&params(&[("n", c["n"].to_string()), ("seed", c["seed"].to_string())])).unwrap();
            let (r, ch) = run_once(&mut inst);
            assert_eq!(r, golden_hex(&c["run"]));
            assert_eq!(ch, golden_hex(&c["check"]));
        }
    }
}
